Report error on unsupported era

locallycompact opened the pull request:
<!-- Describe your change here -->

Required for Conway Support https://github.com/input-output-hk/hydra/issues/1177

⛄ Add e2e tests to check the hydra-node survives a fork (like Conway) and report on unsupported era:
- while running
- on startup: here, because `publishHydraScriptsAs` does not supports the new era Conway, we had to delay the fork so that the hydra scripts are published under the current supported Babbage.

⛄ throw `EraNotSupportedException` in the direct chain component when we detect blocks from a different era.

⛄ Use `displayException` in hydra-node to explain exceptions rised during a change of era, like:
- EraNotSupportedException
- QueryEraMismatchException

> We are also migrating the ConfigurationExceptions to use displayException to 'explain' on the hydra-node 'Run' case.

⛄ Implement helper function (`forkIntoConwayInEpoch`) to simulate a change of era, by using a cardano-config file that will trigger a fork in the cardano-node when it reaches a certain epoch.

⛄ Enhance cardano-node setup in devnet.

⛄ Add missing `genDelegs` attribute to genesis-conway config file, as is needed for cardano-node to run and produce Conway blocks in devnet.

⛄ Reduce `epochLength` in genesis-shelley config file, to trigger a fork in the cardano-node faster.

---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [x] No new TODOs introduced or explained herafter

locallycompact committed:
Add test stub for surviving a conway fork

ch1bo committed:
Draft era mismatch e2e test and configurable Devnet setup

ffakenz committed:
Resolve TODO for setting up cardano node in devnet

ffakenz committed:
Remove unnecessary config from default

ffakenz committed:
Change default cardano node config file name to match ours

As it seems to be confusing otherwise.

ffakenz committed:
Implement forkIntoConwayInEpoch function

This updates the node config file by addin a field
that is going to trigger the fork at certain epoch.

ffakenz committed:
fix forkIntoConwayInEpoch by passing the stateDirectory to find the config file

ffakenz committed:
Add genDelegs to genesis-conway as it is needed for cardano-node to run

locallycompact committed:
Make stderr configurable in withHydraNode

ffakenz committed:
Fetch slotLength and epochLength from args

ffakenz committed:
Introduce ChainClientException and try to catch it as part of e2e

Also change the approach of how to handle the create process, by using alternatives

ch1bo committed:
Fix end-to-end test on non-supported cardano era

ch1bo committed:
Use displayException in hydra-node to explain EraNotSupportedException

This is not yet migrating the ConfigurationExceptions which are relying
on a call to 'explain' on the hydra-node 'Run' case.

ch1bo committed:
Draft another e2e test about node failing on incorrect era

ffakenz committed:
Fix e2e spec to report on startup

By implementing displayException for QueryException.

publishHydraScriptsAs does not supports Conway; for that reason
we delayed the fork so that the hydra scripts are published under Babbage.

ffakenz committed:
Minor formatting changes

ffakenz committed:
Migrate configuration exceptions to use displayException to explain

ffakenz committed:
Update CHANGELOG

ffakenz committed:
Update explain messages for not supported era exceptions

ffakenz committed:
Improve getField helper function by using lenses

ffakenz committed:
Reduce epochLengh even further to speedup tests depending on it

v0d1ch commented on hydra-cluster/src/CardanoNode.hs:
```diff
@@ -405,6 +424,9 @@ data NodeLog
 addField k v = withObject (Aeson.KeyMap.insert k (toJSON v))
 
+getField :: ToJSON a => Aeson.Key -> a -> Aeson.Value
+getField k v = fromMaybe Aeson.Null $ toJSON v ^? (_Object . ix k)
```
You could just do this:
```suggestion
getField k v = fromMaybe Aeson.Null $ toJSON v ^? key k . _Value
```

v0d1ch commented on hydra-node/exe/hydra-node/Main.hs:
```diff
@@ -25,9 +25,9 @@ main = do
   case command of
     Run options ->
-      run (identifyNode options) `catch` (die . explain)
+      run (identifyNode options) `catch` \(SomeException e) -> die $ displayException e
```
Do we need to catch all exceptions here? Perhaps it would be better to use `catches` to specify all exceptions we know that can happen and and provide slightly better user feedback and then have one handler for `SomeException`.

v0d1ch replied:
I see you use `displayException` to do pretty much the same thing :+1:

v0d1ch commented on hydra-node/src/Hydra/Chain/CardanoClient.hs:
```diff
@@ -29,7 +29,11 @@ instance Eq QueryException where
+instance Exception QueryException where
+  displayException = \case
+    QueryAcquireException failure -> show failure
+    QueryEraMismatchException _ -> "Connected to cardano-node in not supported era. Please upgrade your hydra-node."
```
This reads a bit weird to me but @locallycompact would know better 
```suggestion
    QueryEraMismatchException _ -> "Connected to cardano-node in unsupported era. Please upgrade your hydra-node."
```

v0d1ch commented on hydra-node/src/Hydra/Chain/Direct.hs:
```diff
@@ -253,6 +253,14 @@ newtype IntersectionNotFoundException = IntersectionNotFound
+
+instance Exception ChainClientException where
+  displayException = \case
+    EraNotSupportedException -> "Received blocks in not supported era. Please upgrade your hydra-node."
```
ditto

```suggestion
    EraNotSupportedException -> "Received blocks in unsupported era. Please upgrade your hydra-node."
```

v0d1ch commented on hydra-cluster/test/Test/EndToEndSpec.hs:
```diff
@@ -495,6 +511,52 @@ spec = around (showLogsOnFailure "EndToEndSpec") $ do
+delayEpoch stateDirectory args epochs = do
+  shellyGenesisFile :: Aeson.Value <- unsafeDecodeJsonFile (stateDirectory </> nodeShelleyGenesisFile args)
+  let slotLength :: Double =
+        case fromJSON (shellyGenesisFile & getField "slotLength") of
```
Refactor to reduce duplication here?

v0d1ch commented on hydra-cluster/src/HydraNode.hs:
```diff
@@ -388,9 +388,11 @@ withHydraNode' ::
   Maybe Handle ->
-  (Handle -> ProcessHandle -> IO a) ->
+  -- | If given use this as std err.
+  Maybe Handle ->
```
Check if this handle is actually used. I don't see it get set to something other than `Nothing`

v0d1ch reviewed (approved):
Nice work! I added some comments, potential improvements and perhaps one unneeded function parameter.

ffakenz committed:
Apply review changes

ffakenz committed:
Remove stderr to be configurable in withHydraNode, as not needed

ch1bo commented on cabal.project:
```diff
@@ -33,10 +33,6 @@ packages:
--- Warnings as errors for local packages
-program-options
-  ghc-options: -Werror
-
```
Must: not remove this

If you want to turn off `-Werror`, use `cabal configure` (see contributing guidelines).

ch1bo commented on hydra-cluster/test/Test/EndToEndSpec.hs:
```diff
@@ -495,6 +511,46 @@ spec = around (showLogsOnFailure "EndToEndSpec") $ do
+delayEpoch stateDirectory args epochs = do
+  shellyGenesisFile :: Aeson.Value <- unsafeDecodeJsonFile (stateDirectory </> nodeShelleyGenesisFile args)
+  let slotLength :: Double = shellyGenesisFile & getField "slotLength"
+      epochLength = shellyGenesisFile & getField "epochLength"
```
Could: simplify this by using `lens-aeson` (which used by `getField`)

I think this would work here:
```suggestion
      epochLength = shellyGenesisFile ^. key "epochLength"
```

If it does, no need to add `getField`

ch1bo commented on hydra-cluster/src/CardanoNode.hs:
```diff
@@ -241,6 +191,75 @@ withCardanoNodeOnKnownNetwork tracer workDir knownNetwork action = do
+forkIntoConwayInEpoch stateDirectory args n = do
+  config <-
+    unsafeDecodeJsonFile (stateDirectory </> nodeConfigFile args)
+      <&> addField "TestConwayHardForkAtEpoch" n
```
Could: use lens-aeson for this. Found out today that this should be possible here:

```suggestion
      <&> atKey "TestConwayHardForkAtEpoch" ?~ n
```

ch1bo commented on hydra-cluster/src/CardanoNode.hs:
```diff
@@ -241,6 +191,75 @@ withCardanoNodeOnKnownNetwork tracer workDir knownNetwork action = do
+  Aeson.encodeFile
+    (stateDirectory </> nodeConfigFile args)
+    config
+  return ()
```
Could: remove not needed code
```suggestion
```

ch1bo commented on hydra-cluster/src/CardanoNode.hs:
```diff
@@ -241,6 +191,75 @@ withCardanoNodeOnKnownNetwork tracer workDir knownNetwork action = do
+
+-- | Modify the cardano-node configuration to fork into conway at given era
+-- number. See 'config/devnet/genesis-shelley.json' for the epoch length (in
+-- slots).
```
Could: improve haddock (which I wrote) as it is now taking the shelley genesis as given in `CardanoNodeArgs`

ch1bo commented on hydra-cluster/test/Test/EndToEndSpec.hs:
```diff
@@ -539,7 +595,7 @@ timedTx tmpDir tracer node@RunningNode{networkId, nodeSocket} hydraScriptsTxId =
     headId <-
       waitForAllMatch 10 [n1] $
-        headIsInitializingWith (Set.fromList [alice])
+        headIsInitializingWith (Set.fromList [aliceTimed])
```
I also just fixed those today. Instead of renaming them, we can just drop lines `528` and `529` and use the `alice` and `aliceSk` from the fixture.

ch1bo commented on hydra-node/src/Hydra/Chain/CardanoClient.hs:
```diff
@@ -29,7 +29,11 @@ instance Eq QueryException where
+instance Exception QueryException where
+  displayException = \case
+    QueryAcquireException failure -> show failure
+    QueryEraMismatchException _ -> "Connected to cardano-node in unsupported era. Please upgrade your hydra-node."
```
Should: mention the era which is not supported.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct.hs:
```diff
@@ -253,6 +253,14 @@ newtype IntersectionNotFoundException = IntersectionNotFound
+
+instance Exception ChainClientException where
+  displayException = \case
+    EraNotSupportedException -> "Received blocks in unsupported era. Please upgrade your hydra-node."
```
Should: mention which era we saw in the message

ch1bo reviewed (approved):
Change looks good.. but please do revert `cabal.project` before merging.

ch1bo commented on hydra-cluster/test/Test/EndToEndSpec.hs:
```diff
@@ -495,6 +511,46 @@ spec = around (showLogsOnFailure "EndToEndSpec") $ do
+              delayEpoch tmpDir args 1
+
+              waitForProcess ph `shouldReturn` ExitFailure 1
+              hGetContents err >>= (`shouldContain` "Received blocks in not supported era.")
```
Could: reduce the string matching and only assert what we care about. IMO that is `"unsupported era"` and `"upgrade hydra-node"` (maybe via to `shouldContain` checks).

By not matching the exact string, we break this test less likely if we rephrase or extend the message

ffakenz committed:
Revert cabal.project changes in code

If we want to turn off -Werror, we should use cabal configure as
described in the contributing guidelines.

ffakenz committed:
Minor fix to error output assertions

ffakenz committed:
Remove addField helper, use aeson-lenses instead

ffakenz committed:
Improve era missmatch error messages

So they explicitly describe the era missmatch.

Also included some formatting changes.

ffakenz committed:
Enhance error msg for EraNotSupportedException

By providing informationa about the era mismatch that triggered it.

ffakenz committed:
Update fourmolu fixities to match treefmt rules

ffakenz committed:
Minor haddock fix

Remove outdated reference to shelley genesis file from comment.

ffakenz committed:
Remove getField helper, use aeson-lenses instead

ffakenz committed:
Minor formatting changes

ffakenz committed:
Improved delayEpoch so it waits until it reaches the target epoch

Added queryEpochNo to cardano-node client to support this operation.

ffakenz committed:
Revert "Update fourmolu fixities to match treefmt rules"

This reverts commit a8dcedaeb2914830c74b822441c415be64870cd0.

ffakenz committed:
Minor formatting changes

ch1bo committed:
Fix formatting for <&> in presence of ?~

ffakenz committed:
Remove commented code

ffakenz committed:
Improve error messages by using printf
